File: week_01/Java_Code/Subsets.java
import java.util.ArrayList;
import java.util.Arrays;

public class Subsets {
    /**
     * @param S: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    public ArrayList<ArrayList<Integer>> subsets(int[] S) {
    	ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
    	if (S == null || S.length == 0) {
    		return result;
    	}
    	Arrays.sort(S);
    	ArrayList<Integer> list = new ArrayList<Integer>();
    	subsetsHelper(S, result, list, 0);
    	return result;
    }

    private void subsetsHelper(int[] S, ArrayList<ArrayList<Integer>> result,
    	ArrayList<Integer> list, int index) {
    	result.add(new ArrayList<Integer>(list));
    	for (int i = index; i < S.length; i++) {
    		list.add(S[i]);
    		subsetsHelper(S, result, list, i + 1);
    		list.remove(list.size() - 1);
    	}
    }

    public static void main(String[] args) {
    	Subsets test = new Subsets();
    	int[] S = {1, 2, 3};
    	System.out.println(test.subsets(S));
    }
}

File: NineChapters/01.strStr_CodingStyle/Python_code/subsets.py
'''Version 1: my recursion'''
class Solution:
    """
    @param S: The set of numbers.
    @return: A list of lists. See example.
    """
    def subsets(self, S):
        if not S:
            return []
        result = []
        S.sort()
        self.helper(S, [], result, 0)
        return result

    def helper(self, S, lis, result, index):
        if index == len(S):
            result.append(lis)
            return
        self.helper(S, lis, result, index + 1)
        self.helper(S, lis + [S[index]], result, index + 1)

'''Version 2: JiuZhang recursion style'''
class SolutionTwo:
    def subsets(self, S):
        if not S:
            return []
        S.sort()
        result = []
        self.subsets_helper(S, result, [], 0)
        return result

    def subsets_helper(self, S, result, lis, index):
        result.append(lis[:])
        for i in xrange(index, len(S)):
            lis.append(S[i])
            self.subsets_helper(S, result, lis, i + 1)
            lis.pop()

def main():
    test = Solution()
    S = [1,2,3]
    print test.subsets(S)

if __name__ == '__main__':
    main()